core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -17052,36 +17122,40 @@ public Group addGroup(String groupName, List<String> usersIds, List<org.wso2.car
             throw new UserStoreClientException(errorMessage, errorCode);
         }
         Group group;
+        String groupId = generateGroupUUID();
         if (isUniqueGroupIdEnabled(this)) {
             if (!isUniqueUserIdEnabledInUserStore(userStore)) {
                 String errorCode = ErrorMessages.ERROR_CODE_GROUP_UUID_NOT_SUPPORTED.getCode();
                 String errorMessage = ErrorMessages.ERROR_CODE_GROUP_UUID_NOT_SUPPORTED.getMessage();
                 handleAddGroupFailure(errorCode, errorMessage, groupName, null, usersIds, claims);
                 throw new UserStoreException(errorCode + "-" + errorMessage);
             }
-            group = doAddGroup(groupName, usersIds, buildClaimsList(claims));
+            group = doAddGroup(groupName, groupId, usersIds, buildClaimsList(claims));
             groupUniqueIDDomainResolver.setDomainForGroupId(group.getGroupID(), getMyDomainName(), tenantId,
                     false);
         } else {
             // Backward compatibility support. Use group resolver to update the other required places.
             GroupResolver groupResolver = UserStoreMgtDataHolder.getInstance().getGroupResolver();
             try {
-                group = buildGroupFromGroupClaims(groupName, getMyDomainName(), claims);
-                groupResolver.addGroup(group, this);
+                group = groupResolver.addGroup(groupName, groupId, claims, this);
                 if (isUniqueUserIdEnabledInUserStore(userStore)) {
-                    doAddRoleWithID(groupName, usersIds.toArray(new String[0]), false);
+                    doAddGroupWithUserIds(groupName, usersIds);
                 } else {
                     List<User> users = userUniqueIDManger.getUsers(usersIds, this);
-                    doAddRole(groupName, users.stream().map(User::getUsername).toArray(String[]::new), false);
+                    doAddGroupWithUserNames(groupName,
+                            users.stream().map(User::getUsername).collect(Collectors.toList()));
                 }
                 // Update only the cache since the ID can be found in our side.
                 groupUniqueIDDomainResolver.setDomainForGroupId(group.getGroupID(), getMyDomainName(), tenantId,
                         true);
             } catch (UserStoreException e) {
-                groupResolver.deleteGroupByName(groupName, this);
+                log.debug("error occurred while adding group:" + groupName, e);
                 throw e;
+            } finally {
+                groupResolver.deleteGroupByName(groupName, this);
```

Comment:
Was this intentional? Why do we need to execute `deleteGroupByName` inside the finally block here?